Update ctrDetect Binaries

kousu opened the pull request:
Manually version bump ctrDetect to match [upstream](https://www.creatis.insa-lyon.fr/site7/en/ctrDetect)'s distribution.

This should be a minor change. It is in preparation for better automating the build which will download from M. Sdika's upstream directly.

<details><summary>How The Packages Were Built</summary>

```
$ curl -JLO https://osf.io/mka78/?action=download
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   459  100   459    0     0   1577      0 --:--:-- --:--:-- --:--:--  1577
100 74.6M  100 74.6M    0     0  2870k      0  0:00:26  0:00:26 --:--:-- 3842k
curl: Saved to filename '20200420_sct_binaries_linux.tar.gz'
$ curl -JLO https://osf.io/dn67h/?action=download 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   459  100   459    0     0   1645      0 --:--:-- --:--:-- --:--:--  1645
100 57.8M  100 57.8M    0     0  3473k      0  0:00:17  0:00:17 --:--:-- 3851k
curl: Saved to filename '20200420_sct_binaries_osx.tar.gz'
```

Follow https://github.com/neuropoly/spinalcordtoolbox/wiki/Compile-binaries -> https://www.creatis.insa-lyon.fr/site7/en/ctrDetect ->

```
$ curl -JLO http://www.creatis.insa-lyon.fr/~sdika/soft/ctrDetect-v1_x86_64.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   279  100   279    0     0    690      0 --:--:-- --:--:-- --:--:--   692
100 33.9M  100 33.9M    0     0  1092k      0  0:00:31  0:00:31 --:--:-- 1131k
$ curl -JLO http://www.creatis.insa-lyon.fr/~sdika/soft/ctrDetect-v1_macos10.11.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   283  100   283    0     0    686      0 --:--:-- --:--:-- --:--:--   685
100  9.8M  100  9.8M    0     0  1966k      0  0:00:05  0:00:05 --:--:-- 2520k
```

Unpack the old version, update ctrDetect, and repack:

```
$ mkdir binaries_osx; tar -zxvf 20200420_sct_binaries_osx.tar.gz -C binaries_osx
./
./isct_dice_coefficient
./isct_antsRegistration
./isct_spine_detect
./isct_ComposeMultiTransform
./isct_train_svm
./isct_propseg
./isct_antsSliceRegularizedRegistration
./isct_antsApplyTransforms
$ tar -zxvf ctrDetect-v1_macos10.11.tar.gz 
ctrDetect/
ctrDetect/LICENSE.txt
ctrDetect/LICENSE_opencv.txt
ctrDetect/README.txt
ctrDetect/README.txt~
ctrDetect/classif_CC.yml
ctrDetect/rectifyCtr.py
ctrDetect/spine_detect
ctrDetect/spine_detect.bash
ctrDetect/spine_train_svm
ctrDetect/spine_train_svm.bash
$ cp -p ctrDetect/spine_detect binaries_osx/isct_spine_detect  # -p to preserve the timestamp
$ cp -p ctrDetect/spine_train_svm binaries_osx/isct_train_svm
$ ls -l binaries_osx/
total 197284
-rwxr-xr-x 1 kousu kousu 32203680 Apr 19 14:44 isct_antsApplyTransforms
-rwxr-xr-x 1 kousu kousu 48876288 Apr 19 14:44 isct_antsRegistration
-rwxr-xr-x 1 kousu kousu 24320568 Apr 19 14:44 isct_antsSliceRegularizedRegistration
-rwxr-xr-x 1 kousu kousu 22713500 Apr 19 14:44 isct_ComposeMultiTransform
-rwxr-xr-x 1 kousu kousu  9589820 Mar  2  2016 isct_dice_coefficient
-rwxr-xr-x 1 kousu kousu 36514576 Sep 18  2016 isct_propseg
-rwxr-xr-x 1 kousu kousu 13873744 Apr 11  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 13909424 Apr 11  2017 isct_train_svm
$ tar -czvf 20200421_sct_binaries_osx.tar.gz -C binaries_osx/ .
./
./isct_antsRegistration
./isct_propseg
./isct_train_svm
./isct_dice_coefficient
./isct_ComposeMultiTransform
./isct_antsApplyTransforms
./isct_spine_detect
./isct_antsSliceRegularizedRegistration
```

Now do the Linux version:

```
$ mkdir binaries_linux; tar -zxvf 20200420_sct_binaries_linux.tar.gz -C binaries_linux
./
./isct_dice_coefficient
./isct_antsRegistration
./isct_spine_detect
./isct_ComposeMultiTransform
./isct_train_svm
./isct_propseg
./isct_antsSliceRegularizedRegistration
./isct_antsApplyTransforms   
$ tar -zxvf ctrDetect-v1_x86_64.tar.gz 
ctrDetect/
ctrDetect/libc.so.6
ctrDetect/classif_CC.yml
ctrDetect/README.txt~
ctrDetect/LICENSE_opencv.txt
ctrDetect/spine_detect.bash
ctrDetect/spine_detect
ctrDetect/spine_train_svm.bash
ctrDetect/libc.so
ctrDetect/README.txt
ctrDetect/spine_train_svm
ctrDetect/LICENSE.txt
$ ls -l binaries_linux/
total 246112
-rwxr-xr-x 1 kousu kousu 28630112 Apr 19 15:08 isct_antsApplyTransforms
-rwxr-xr-x 1 kousu kousu 44594624 Apr 19 15:08 isct_antsRegistration
-rwxr-xr-x 1 kousu kousu 21435952 Apr 19 15:08 isct_antsSliceRegularizedRegistration
-rwxr-xr-x 1 kousu kousu 19880216 Apr 19 15:08 isct_ComposeMultiTransform
-rwxr-xr-x 1 kousu kousu 10439872 Mar  2  2016 isct_dice_coefficient
-rwxr-xr-x 1 kousu kousu 20504754 Jul  5  2016 isct_propseg
-rwxr-xr-x 1 kousu kousu 53186384 Apr  9  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 53331105 Dec  4  2018 isct_train_svm
$ cp -p ctrDetect/spine_detect binaries_linux/isct_spine_detect 
$ cp -p ctrDetect/spine_train_svm binaries_linux/isct_train_svm 
$ ls -l binaries_linux/
total 246100
-rwxr-xr-x 1 kousu kousu 28630112 Apr 19 15:08 isct_antsApplyTransforms
-rwxr-xr-x 1 kousu kousu 44594624 Apr 19 15:08 isct_antsRegistration
-rwxr-xr-x 1 kousu kousu 21435952 Apr 19 15:08 isct_antsSliceRegularizedRegistration
-rwxr-xr-x 1 kousu kousu 19880216 Apr 19 15:08 isct_ComposeMultiTransform
-rwxr-xr-x 1 kousu kousu 10439872 Mar  2  2016 isct_dice_coefficient
-rwxr-xr-x 1 kousu kousu 20504754 Jul  5  2016 isct_propseg
-rwxr-xr-x 1 kousu kousu 53172736 Apr 19  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 53331105 Apr 19  2017 isct_train_svm
$ tar -czvf 20200421_sct_binaries_linux.tar.gz -C binaries_linux/ .
./
./isct_antsRegistration
./isct_propseg
./isct_train_svm
./isct_dice_coefficient
./isct_ComposeMultiTransform
./isct_antsApplyTransforms
./isct_spine_detect
./isct_antsSliceRegularizedRegistration 
```

Checksums:

```
$ sha256sum 20200421_sct_binaries_*
e1968ddf15094f860777cbe2167870e2100662db951da78fcb79f8bf6190c394  20200421_sct_binaries_linux.tar.gz
b3365d0e125a7f9f5bd1726acfb75641b3e1c33e50acfc9ac4bb274cbaea5fc2  20200421_sct_binaries_osx.tar.gz
```

</details>

kousu commented:
Ah, that's weird. This actually *downgraded* ctrDetect, if I can trust the timestamps. The OS X version we were distributing was built a day after the version [from here](http://www.creatis.insa-lyon.fr/~sdika/soft/ctrDetect-v1_macos10.11.tar.gz), and the Linux version we were distributing 

Re-extract the ctrDetect binaries, old and new and for both platforms:

```
$ mkdir 20200420_sct_binaries_linux; tar -C 20200420_sct_binaries_linux -zxvf 20200420_sct_binaries_linux.tar.gz ./isct_train_svm ./isct_spine_detects
./isct_spine_detect
./isct_train_svm
$ mkdir 20200421_sct_binaries_linux; tar -C 20200421_sct_binaries_linux -zxvf 20200421_sct_binaries_linux.tar.gz ./isct_train_svm ./isct_spine_detect
./isct_train_svm
./isct_spine_detect
$ mkdir 20200420_sct_binaries_osx; tar -C 20200420_sct_binaries_osx -zxvf 20200420_sct_binaries_osx.tar.gz ./isct_train_svm ./isct_spine_detect
./isct_spine_detect
./isct_train_svm
$ mkdir 20200421_sct_binaries_osx; tar -C 20200421_sct_binaries_osx -zxvf 20200421_sct_binaries_osx.tar.gz ./isct_train_svm ./isct_spine_detect
./isct_train_svm
./isct_spine_detect
```

Compare dates:

```
$ ls -l 2020042*_sct_binaries_linux
20200420_sct_binaries_linux:
total 104028
-rwxr-xr-x 1 kousu kousu 53186384 Apr  9  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 53331105 Dec  4  2018 isct_train_svm

20200421_sct_binaries_linux:
total 104012
-rwxr-xr-x 1 kousu kousu 53172736 Apr 19  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 53331105 Apr 19  2017 isct_train_svm
$ ls -l 2020042*_sct_binaries_osx
20200420_sct_binaries_osx:
total 27504
-rwxr-xr-x 1 kousu kousu 14071264 Apr 12  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 14090052 Apr 12  2017 isct_train_svm

20200421_sct_binaries_osx:
total 27136
-rwxr-xr-x 1 kousu kousu 13873744 Apr 11  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 13909424 Apr 11  2017 isct_train_svm
```

The new Linux dist has a newer `spine_detect` -- by 10 days -- but an older `train_svm` -- by a year and a half (!).

The new OS X dist is actually a day older than what we've been distributing.

I tried looking in [the other available OS X dist](http://www.creatis.insa-lyon.fr/~sdika/soft/ctrDetect-v1_macos10.7.tar.gz) for comparison

```
20200420_sct_binaries_osx.tar.gz    20200421_sct_binaries_osx.tar.gz    ctrDetect-v1_x86_64.tar.gz
$ tar -zxvf ctrDetect-v1_macos10.11.tar.gz 
ctrDetect/
ctrDetect/LICENSE.txt
ctrDetect/LICENSE_opencv.txt
ctrDetect/README.txt
ctrDetect/README.txt~
ctrDetect/classif_CC.yml
ctrDetect/rectifyCtr.py
ctrDetect/spine_detect
ctrDetect/spine_detect.bash
ctrDetect/spine_train_svm
ctrDetect/spine_train_svm.bash
$ tar -zxvf ctrDetect-v1_macos10.7.tar.gz 
ctrDetect10.7/
ctrDetect10.7/LICENSE.txt
ctrDetect10.7/LICENSE_opencv.txt
ctrDetect10.7/README.txt
ctrDetect10.7/classif_CC.yml
ctrDetect10.7/rectifyCtr.py
ctrDetect10.7/spine_detect
ctrDetect10.7/spine_detect.bash
ctrDetect10.7/spine_train_svm
ctrDetect10.7/spine_train_svm.bash

$ sha256sum 20200420_sct_binaries_osx/*train_svm ctrDetect*/*train_svm
f1bc1ceb4db2ec2ab046800648d5dcc7eb547736bd3559e6e9c17ed44f1ed1f8  20200420_sct_binaries_osx/isct_train_svm
58df50b57d67243af92f10fbfd079ff35f32081fd150866a36424e46592c18cd  ctrDetect10.7/spine_train_svm
1ade423edf0ef085edabea1ba33b900b92f22309bef78443ed06293cb07dce1e  ctrDetect/spine_train_svm
$ sha256sum 20200420_sct_binaries_osx/*spine_detect ctrDetect*/*spine_detect
07d3f91562878ad9b3375385ebcf093d0ae72d399427dbc7dabd5795c0e4374c  20200420_sct_binaries_osx/isct_spine_detect
d2c11bde2a0c4bb77b1eddcc52917167589ef78ffd216c81684c3798fa6db379  ctrDetect10.7/spine_detect
004b430bd3646d591aa236699f446e42bc974974db36f580d73c1795e9af9d13  ctrDetect/spine_detect
```

So what we have been distributing is _neither_ what M. Sdika has been distributing. Where did these binaries come from? Why is there one file we have that was recompiled a year and a half later that never got updated in the upstream distribution?

kousu commented:
I've rebased and rebuilt the binary dependency distributions to match the updates I made to #2642.

<details><summary>Build log</summary>

Working from `jca/947-new-ants-binaries`, I downloaded that version's binaries and merged in the binaries directly from [M. Sdika](https://www.creatis.insa-lyon.fr/site7/en/ctrDetect) again.

## Linux

```
(sct_venv) $ sct_download_data -d binaries_linux -o bin.new-ants-binaries.linux

--
Spinal Cord Toolbox (git-jca/947-new-ants-binaries-77b2a3bc57dae061b6cda68fea6b0276fb320a20)


Trying URL: https://osf.io/mka78/?action=download
Downloading 20200420_sct_binaries_linux.tar.gz...
Status: 100%|################################################################################################################################################| 78.3M/78.3M [00:19<00:00, 3.92MB/s]

Create temporary folder (/tmp/sct-20200426152704.354181-p9_vsvhz)...

Unzip data to: /tmp/sct-20200426152704.354181-p9_vsvhz

Check if folder already exists on the destination path...

Remove temporary folders...
Done!

(sct_venv) $ ls -l bin.new-ants-binaries.linux/
total 246396
-rwxr-xr-x 1 kousu kousu 28630112 Apr 22 16:11 isct_antsApplyTransforms
-rwxr-xr-x 1 kousu kousu 44594624 Apr 22 16:11 isct_antsRegistration
-rwxr-xr-x 1 kousu kousu 21435952 Apr 22 16:11 isct_antsSliceRegularizedRegistration
-rwxr-xr-x 1 kousu kousu 19880216 Apr 22 16:11 isct_ComposeMultiTransform
-rwxr-xr-x 1 kousu kousu 10439872 Mar  2  2016 isct_dice_coefficient
-rwxr-xr-x 1 kousu kousu 20504754 Jul  5  2016 isct_propseg
-rwxr-xr-x 1 kousu kousu 53186384 Apr  9  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 53331105 Dec  4  2018 isct_train_svm
(sct_venv) $ sha256sum bin.new-ants-binaries.linux/*
f70984d34f0d40af327d01ac3bcacd4ec9029b3d96b79008c1ef76dccac52ca7  bin.new-ants-binaries.linux/isct_antsApplyTransforms
8bd0ee2fb22051e96d6c00e09dcce24777e34ece0b13e7a09e666e38bbbda00b  bin.new-ants-binaries.linux/isct_antsRegistration
49896d49e4bdfcdc0edb9d70c2c87e6508b16ebc360c82e5515604a55a856af8  bin.new-ants-binaries.linux/isct_antsSliceRegularizedRegistration
bde2c9689352f8b73f38b5ca7c9647ce0057ee3479eac6e221c926fa3d490ce1  bin.new-ants-binaries.linux/isct_ComposeMultiTransform
58d0487d37d7c994f2a6fa3820d715de4852f63d554be9f0177ebb900234f623  bin.new-ants-binaries.linux/isct_dice_coefficient
270fb99132d91c3f890caec28efbb8c314857a9e4c843ddd9c7fab3dcb241218  bin.new-ants-binaries.linux/isct_propseg
d8dd029bb9c76d979e2edacbcf04e44d34e5c10d7efdabd2831a11f3bdd903be  bin.new-ants-binaries.linux/isct_spine_detect
699d42b6dd10430af664f63fd1d2105a8a42c46bae83f86e082435f3932cc439  bin.new-ants-binaries.linux/isct_train_svm
(sct_venv) $ curl -JLO http://www.creatis.insa-lyon.fr/~sdika/soft/ctrDetect-v1_x86_64.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   279  100   279    0     0    747      0 --:--:-- --:--:-- --:--:--   745
100 33.9M  100 33.9M    0     0  2551k      0  0:00:13  0:00:13 --:--:-- 2567k
(sct_venv) $ tar -zxvf ctrDetect-v1_x86_64.tar.gz 
ctrDetect/
ctrDetect/libc.so.6
ctrDetect/classif_CC.yml
ctrDetect/README.txt~
ctrDetect/LICENSE_opencv.txt
ctrDetect/spine_detect.bash
ctrDetect/spine_detect
ctrDetect/spine_train_svm.bash
ctrDetect/libc.so
ctrDetect/README.txt
ctrDetect/spine_train_svm
ctrDetect/LICENSE.txt

(sct_venv) $ cp -rp bin.new-ants-binaries.linux/ bin.new-ctrDetect-binaries.linux/
(sct_venv) $ cp -p ctrDetect/spine_detect bin.new-ctrDetect-binaries.linux/isct_spine_detect 
(sct_venv) $ cp -p ctrDetect/spine_train_svm bin.new-ctrDetect-binaries.linux/isct_train_svm 

(sct_venv) $ ls -l bin.new-ctrDetect-binaries.linux/
total 246384
-rwxr-xr-x 1 kousu kousu 28630112 Apr 22 16:11 isct_antsApplyTransforms
-rwxr-xr-x 1 kousu kousu 44594624 Apr 22 16:11 isct_antsRegistration
-rwxr-xr-x 1 kousu kousu 21435952 Apr 22 16:11 isct_antsSliceRegularizedRegistration
-rwxr-xr-x 1 kousu kousu 19880216 Apr 22 16:11 isct_ComposeMultiTransform
-rwxr-xr-x 1 kousu kousu 10439872 Mar  2  2016 isct_dice_coefficient
-rwxr-xr-x 1 kousu kousu 20504754 Jul  5  2016 isct_propseg
-rwxr-xr-x 1 kousu kousu 53172736 Apr 19  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 53331105 Apr 19  2017 isct_train_svm
(sct_venv) $ sha256sum bin.new-ctrDetect-binaries.linux/*
f70984d34f0d40af327d01ac3bcacd4ec9029b3d96b79008c1ef76dccac52ca7  bin.new-ctrDetect-binaries.linux/isct_antsApplyTransforms
8bd0ee2fb22051e96d6c00e09dcce24777e34ece0b13e7a09e666e38bbbda00b  bin.new-ctrDetect-binaries.linux/isct_antsRegistration
49896d49e4bdfcdc0edb9d70c2c87e6508b16ebc360c82e5515604a55a856af8  bin.new-ctrDetect-binaries.linux/isct_antsSliceRegularizedRegistration
bde2c9689352f8b73f38b5ca7c9647ce0057ee3479eac6e221c926fa3d490ce1  bin.new-ctrDetect-binaries.linux/isct_ComposeMultiTransform
58d0487d37d7c994f2a6fa3820d715de4852f63d554be9f0177ebb900234f623  bin.new-ctrDetect-binaries.linux/isct_dice_coefficient
270fb99132d91c3f890caec28efbb8c314857a9e4c843ddd9c7fab3dcb241218  bin.new-ctrDetect-binaries.linux/isct_propseg
7d14eae7b0ab6eb3751f347f078e7147b4d672f568afd5870d0c11ab7b7d05c9  bin.new-ctrDetect-binaries.linux/isct_spine_detect
699d42b6dd10430af664f63fd1d2105a8a42c46bae83f86e082435f3932cc439  bin.new-ctrDetect-binaries.linux/isct_train_svm

(sct_venv) $ rm -rf ctrDetect

(sct_venv) $ tar --sort=name -czvf 20200421_sct_binaries_linux.tar.gz -C bin.new-ctrDetect-binaries.linux/ .
./
./isct_ComposeMultiTransform
./isct_antsApplyTransforms
./isct_antsRegistration
./isct_antsSliceRegularizedRegistration
./isct_dice_coefficient
./isct_propseg
./isct_spine_detect
./isct_train_svm
(sct_venv) $ sha256sum 20200421_sct_binaries_linux.tar.gz 
7d49be92ce72df59d1b4bb6578f28d00478b9fd8b9e2000e8ed038d8d127325b  20200421_sct_binaries_linux.tar.gz
```

## OS X


```
(sct_venv) $ # unpack old version
(sct_venv) $ sct_download_data -d binaries_osx -o bin.new-ants-binaries.osx

--
Spinal Cord Toolbox (git-ko/new-ctrDetect-binaries-77b2a3bc57dae061b6cda68fea6b0276fb320a20)


Trying URL: https://osf.io/dn67h/?action=download
Downloading 20200420_sct_binaries_osx.tar.gz...
Status: 100%|################################################################################################################################################| 60.7M/60.7M [00:14<00:00, 4.08MB/s]

Create temporary folder (/tmp/sct-20200426155240.481763-bspj__9r)...

Unzip data to: /tmp/sct-20200426155240.481763-bspj__9r

Check if folder already exists on the destination path...

Remove temporary folders...
Done!

(sct_venv) $ curl -JLO http://www.creatis.insa-lyon.fr/~sdika/soft/ctrDetect-v1_macos10.11.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   283  100   283    0     0   1164      0 --:--:-- --:--:-- --:--:--  1164
100  9.8M  100  9.8M    0     0  2510k      0  0:00:04  0:00:04 --:--:-- 2675k
(sct_venv) $ tar -zxf ctrDetect-v1_macos10.11.tar.gz
(sct_venv) $ 
(sct_venv) $ # checksum old version
(sct_venv) $ ls -l bin.new-ants-binaries.osx
total 197876
-rwxr-xr-x 1 kousu kousu 32204628 Apr 22 15:54 isct_antsApplyTransforms
-rwxr-xr-x 1 kousu kousu 48869124 Apr 22 15:54 isct_antsRegistration
-rwxr-xr-x 1 kousu kousu 24317428 Apr 22 15:54 isct_antsSliceRegularizedRegistration
-rwxr-xr-x 1 kousu kousu 22710352 Apr 22 15:54 isct_ComposeMultiTransform
-rwxr-xr-x 1 kousu kousu  9589820 Mar  2  2016 isct_dice_coefficient
-rwxr-xr-x 1 kousu kousu 36514576 Sep 18  2016 isct_propseg
-rwxr-xr-x 1 kousu kousu 14071264 Apr 12  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 14090052 Apr 12  2017 isct_train_svm
(sct_venv) $ sha256sum bin.new-ants-binaries.osx/*
dce890134f54fd9288a7d29bb07821210a9098ad61d233cb5969fcb68f76dcd1  bin.new-ants-binaries.osx/isct_antsApplyTransforms
28cf7550acc47de475513587abdbe5fb28c937d76f20e44b305848e1736505ff  bin.new-ants-binaries.osx/isct_antsRegistration
6f9e5b1726f2e00bc310b37bd24806d1d1bbff14d68ec56077d3459e95667670  bin.new-ants-binaries.osx/isct_antsSliceRegularizedRegistration
df88888699985e6575ddc968c630fe6b379dd7a2bb3b6573b97afb95dec2fa57  bin.new-ants-binaries.osx/isct_ComposeMultiTransform
83814cf371e053aa13f941327617ea298a1d4a4f9a8761a5c6a3d08adb7e64b0  bin.new-ants-binaries.osx/isct_dice_coefficient
10c842767d4a81c454a9ed08ac74dd1ad2331cffca71a3acc5d42ddded389afc  bin.new-ants-binaries.osx/isct_propseg
07d3f91562878ad9b3375385ebcf093d0ae72d399427dbc7dabd5795c0e4374c  bin.new-ants-binaries.osx/isct_spine_detect
f1bc1ceb4db2ec2ab046800648d5dcc7eb547736bd3559e6e9c17ed44f1ed1f8  bin.new-ants-binaries.osx/isct_train_svm
(sct_venv) $ 
(sct_venv) $ # update
(sct_venv) $ cp -rp bin.new-ants-binaries.osx bin.new-ctrDetect-binaries.osx
(sct_venv) $ cp -p ctrDetect/spine_detect bin.new-ctrDetect-binaries.osx/isct_spine_detect 
(sct_venv) $ cp -p ctrDetect/spine_train_svm bin.new-ctrDetect-binaries.osx/isct_train_svm 
(sct_venv) $ 
(sct_venv) $ # checksum new version
(sct_venv) $ ls -l bin.new-ctrDetect-binaries.osx
total 197508
-rwxr-xr-x 1 kousu kousu 32204628 Apr 22 15:54 isct_antsApplyTransforms
-rwxr-xr-x 1 kousu kousu 48869124 Apr 22 15:54 isct_antsRegistration
-rwxr-xr-x 1 kousu kousu 24317428 Apr 22 15:54 isct_antsSliceRegularizedRegistration
-rwxr-xr-x 1 kousu kousu 22710352 Apr 22 15:54 isct_ComposeMultiTransform
-rwxr-xr-x 1 kousu kousu  9589820 Mar  2  2016 isct_dice_coefficient
-rwxr-xr-x 1 kousu kousu 36514576 Sep 18  2016 isct_propseg
-rwxr-xr-x 1 kousu kousu 13873744 Apr 11  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 13909424 Apr 11  2017 isct_train_svm
(sct_venv) $ sha256sum bin.new-ctrDetect-binaries.osx/*
dce890134f54fd9288a7d29bb07821210a9098ad61d233cb5969fcb68f76dcd1  bin.new-ctrDetect-binaries.osx/isct_antsApplyTransforms
28cf7550acc47de475513587abdbe5fb28c937d76f20e44b305848e1736505ff  bin.new-ctrDetect-binaries.osx/isct_antsRegistration
6f9e5b1726f2e00bc310b37bd24806d1d1bbff14d68ec56077d3459e95667670  bin.new-ctrDetect-binaries.osx/isct_antsSliceRegularizedRegistration
df88888699985e6575ddc968c630fe6b379dd7a2bb3b6573b97afb95dec2fa57  bin.new-ctrDetect-binaries.osx/isct_ComposeMultiTransform
83814cf371e053aa13f941327617ea298a1d4a4f9a8761a5c6a3d08adb7e64b0  bin.new-ctrDetect-binaries.osx/isct_dice_coefficient
10c842767d4a81c454a9ed08ac74dd1ad2331cffca71a3acc5d42ddded389afc  bin.new-ctrDetect-binaries.osx/isct_propseg
004b430bd3646d591aa236699f446e42bc974974db36f580d73c1795e9af9d13  bin.new-ctrDetect-binaries.osx/isct_spine_detect
1ade423edf0ef085edabea1ba33b900b92f22309bef78443ed06293cb07dce1e  bin.new-ctrDetect-binaries.osx/isct_train_svm
(sct_venv) $ 
(sct_venv) $ # repack
(sct_venv) $ tar --sort=name -czvf 20200421_sct_binaries_osx.tar.gz -C bin.new-ctrDetect-binaries.osx/ .
./
./isct_ComposeMultiTransform
./isct_antsApplyTransforms
./isct_antsRegistration
./isct_antsSliceRegularizedRegistration
./isct_dice_coefficient
./isct_propseg
./isct_spine_detect
./isct_train_svm
(sct_venv) $ sha256sum 20200421_sct_binaries_osx.tar.gz
87f535fd2cbb8b83ac1c9084b5046e1f2f4f36f0ea3958505a3602e468398dd5  20200421_sct_binaries_osx.tar.gz
(sct_venv) $ rm -rf ctrDetect
(sct_venv) $ 
```

</details>

Interestingly, in the Linux version, though `isct_train_svm` had its timestamp jump backwards by a year and a half *the contents are identical*; only `isct_spine_detect` actually changed, and its timestamp says it jumped forward by 10 days, so it couldn't have that many differences. The OS X files are both different, but only by a day.

```
(sct_venv) $ diff -ru bin.new-*.linux
Binary files bin.new-ants-binaries.linux/isct_spine_detect and bin.new-ctrDetect-binaries.linux/isct_spine_detect differ
(sct_venv) $ diff -ru bin.new-*.osx
Binary files bin.new-ants-binaries.osx/isct_spine_detect and bin.new-ctrDetect-binaries.osx/isct_spine_detect differ
Binary files bin.new-ants-binaries.osx/isct_train_svm and bin.new-ctrDetect-binaries.osx/isct_train_svm differ
```

I guess M. Sdika recompiled these for us at some point for some reason? Do we know why?

jcohenadad commented:
> I guess M. Sdika recompiled these for us at some point for some reason? Do we know why?

I'm quickly glancing at my emails but nothing obvious comes up-- i could spend more time but maybe @charleygros has a quick answer?

kousu commented:
> > I guess M. Sdika recompiled these for us at some point for some reason? Do we know why?
> 
> I'm quickly glancing at my emails but nothing obvious comes up-- i could spend more time but maybe @charleygros has a quick answer?

The upstream `ctrDetect` and the ones we've been distributing have subtly different math: https://travis-ci.org/github/neuropoly/spinalcordtoolbox/jobs/679825255 failed with 

```
   Motion parameters do not match:   results: [0.008491259664160821, 0.00290239743577736, 0.00290239743577736, 0.00290239743577736, -0.014405747222058809, -0.014405747222058809, -0.014405747222058809]  ground truth: [0.008491259664160821, 0.0029020670607778237, 0.0029020670607778237, 0.0029020670607778237, -0.014405996135095476, -0.014405996135095476, -0.014405996135095476]
```

It's more stable than before 02f72a91f97dee8f6a52877877995fddd5a26969, only showing differences at 1e-7 and [rest of the builds](https://travis-ci.org/github/neuropoly/spinalcordtoolbox/builds/679825250) passed by chance, presumably the instability was smaller than [`allclose()`](https://numpy.org/doc/1.18/reference/generated/numpy.allclose.html)'s default tolerance.

I'm going to do the same [exhaustive testing](https://github.com/neuropoly/spinalcordtoolbox/pull/2642#issuecomment-619385622) I just did to see just how unstable this makes things.

kousu commented:
<details><summary>Test setup</summary>

```
$ git fetch
$ rm -rf bin
$ 
$ git checkout jca/947-new-ants-binaries && sct_download_data -d binaries_linux -o bin.new-ants-binaries.linux
Switched to branch 'jca/947-new-ants-binaries'
Your branch is up to date with 'origin/jca/947-new-ants-binaries'.

--
Spinal Cord Toolbox (git-jca/947-new-ants-binaries-77b2a3bc57dae061b6cda68fea6b0276fb320a20)


Trying URL: https://osf.io/mka78/?action=download
Downloading 20200420_sct_binaries_linux.tar.gz...
Status: 100%|####################################################################################################| 78.3M/78.3M [00:20<00:00, 3.74MB/s]

Create temporary folder (/tmp/sct-20200426180826.206367-xgpwx5h1)...

Unzip data to: /tmp/sct-20200426180826.206367-xgpwx5h1

Check if folder already exists on the destination path...

Remove temporary folders...
Done!

$ git checkout ko/new-ctrDetect-binaries && sct_download_data -d binaries_linux -o bin.new-ctrDetect-binaries.linux
Switched to branch 'ko/new-ctrDetect-binaries'
Your branch is up to date with 'origin/ko/new-ctrDetect-binaries'.

--
Spinal Cord Toolbox (git-ko/new-ctrDetect-binaries-5e5d16285fc3fcf41d7f9551eb05284845049eec)


Trying URL: https://osf.io/admxy/?action=download
Downloading 20200421_sct_binaries_linux.tar.gz...
Status: 100%|####################################################################################################| 78.2M/78.2M [00:19<00:00, 4.07MB/s]

Create temporary folder (/tmp/sct-20200426182741.006422-9iryxcgz)...

Unzip data to: /tmp/sct-20200426182741.006422-9iryxcgz

Check if folder already exists on the destination path...

Remove temporary folders...
Done!

```

<details><summary>Timestamps + Checksums</summary>

```
$ ls -l bin.new-ants-binaries.linux/
total 246124
-rwxr-xr-x 1 kousu kousu 28630112 Apr 22 16:11 isct_antsApplyTransforms
-rwxr-xr-x 1 kousu kousu 44594624 Apr 22 16:11 isct_antsRegistration
-rwxr-xr-x 1 kousu kousu 21435952 Apr 22 16:11 isct_antsSliceRegularizedRegistration
-rwxr-xr-x 1 kousu kousu 19880216 Apr 22 16:11 isct_ComposeMultiTransform
-rwxr-xr-x 1 kousu kousu 10439872 Mar  2  2016 isct_dice_coefficient
-rwxr-xr-x 1 kousu kousu 20504754 Jul  5  2016 isct_propseg
-rwxr-xr-x 1 kousu kousu 53186384 Apr  9  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 53331105 Dec  4  2018 isct_train_svm
$ sha256sum bin.new-ants-binaries.linux/
sha256sum: bin.new-ants-binaries.linux/: Is a directory
$ sha256sum bin.new-ants-binaries.linux/*
f70984d34f0d40af327d01ac3bcacd4ec9029b3d96b79008c1ef76dccac52ca7  bin.new-ants-binaries.linux/isct_antsApplyTransforms
8bd0ee2fb22051e96d6c00e09dcce24777e34ece0b13e7a09e666e38bbbda00b  bin.new-ants-binaries.linux/isct_antsRegistration
49896d49e4bdfcdc0edb9d70c2c87e6508b16ebc360c82e5515604a55a856af8  bin.new-ants-binaries.linux/isct_antsSliceRegularizedRegistration
bde2c9689352f8b73f38b5ca7c9647ce0057ee3479eac6e221c926fa3d490ce1  bin.new-ants-binaries.linux/isct_ComposeMultiTransform
58d0487d37d7c994f2a6fa3820d715de4852f63d554be9f0177ebb900234f623  bin.new-ants-binaries.linux/isct_dice_coefficient
270fb99132d91c3f890caec28efbb8c314857a9e4c843ddd9c7fab3dcb241218  bin.new-ants-binaries.linux/isct_propseg
d8dd029bb9c76d979e2edacbcf04e44d34e5c10d7efdabd2831a11f3bdd903be  bin.new-ants-binaries.linux/isct_spine_detect
699d42b6dd10430af664f63fd1d2105a8a42c46bae83f86e082435f3932cc439  bin.new-ants-binaries.linux/isct_train_svm
$ ls -l bin.new-ctrDetect-binaries.linux/
total 246112
-rwxr-xr-x 1 kousu kousu 28630112 Apr 22 16:11 isct_antsApplyTransforms
-rwxr-xr-x 1 kousu kousu 44594624 Apr 22 16:11 isct_antsRegistration
-rwxr-xr-x 1 kousu kousu 21435952 Apr 22 16:11 isct_antsSliceRegularizedRegistration
-rwxr-xr-x 1 kousu kousu 19880216 Apr 22 16:11 isct_ComposeMultiTransform
-rwxr-xr-x 1 kousu kousu 10439872 Mar  2  2016 isct_dice_coefficient
-rwxr-xr-x 1 kousu kousu 20504754 Jul  5  2016 isct_propseg
-rwxr-xr-x 1 kousu kousu 53172736 Apr 19  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 53331105 Apr 19  2017 isct_train_svm
$ sha2 bin.new-ctrDetect-binaries.linux/
sha224sum  sha256sum  
$ sha256sum bin.new-ctrDetect-binaries.linux/*
f70984d34f0d40af327d01ac3bcacd4ec9029b3d96b79008c1ef76dccac52ca7  bin.new-ctrDetect-binaries.linux/isct_antsApplyTransforms
8bd0ee2fb22051e96d6c00e09dcce24777e34ece0b13e7a09e666e38bbbda00b  bin.new-ctrDetect-binaries.linux/isct_antsRegistration
49896d49e4bdfcdc0edb9d70c2c87e6508b16ebc360c82e5515604a55a856af8  bin.new-ctrDetect-binaries.linux/isct_antsSliceRegularizedRegistration
bde2c9689352f8b73f38b5ca7c9647ce0057ee3479eac6e221c926fa3d490ce1  bin.new-ctrDetect-binaries.linux/isct_ComposeMultiTransform
58d0487d37d7c994f2a6fa3820d715de4852f63d554be9f0177ebb900234f623  bin.new-ctrDetect-binaries.linux/isct_dice_coefficient
270fb99132d91c3f890caec28efbb8c314857a9e4c843ddd9c7fab3dcb241218  bin.new-ctrDetect-binaries.linux/isct_propseg
7d14eae7b0ab6eb3751f347f078e7147b4d672f568afd5870d0c11ab7b7d05c9  bin.new-ctrDetect-binaries.linux/isct_spine_detect
699d42b6dd10430af664f63fd1d2105a8a42c46bae83f86e082435f3932cc439  bin.new-ctrDetect-binaries.linux/isct_train_svm
```

</details>

<details><summary>Test log</summary>

```
$ for branch in jca/947-new-ants-binaries ko/new-ctrDetect-binaries; do
>   git checkout $branch || break
>   for binaries in bin.new-ants-binaries.linux bin.new-ctrDetect-binaries.linux; do
>     ln -sfT $binaries bin
>     echo
>     echo "----------$branch $binaries-----------"
>     ls -l bin/ # verify we're using the expected binaries
>     date # timestamp
>     for i in `seq 100`; do
>       echo `(cd sct_testing_data; sct_dmri_moco -i dmri/dmri.nii.gz -bvec dmri/bvecs.txt -g 3 -x nn -ofolder dmri_test1 -r 1 >/dev/null 2>&1; cat dmri_test1/moco_params.tsv | cut -f 1)`
>     done |
>     sort | uniq -c | tee $(basename $branch)-bin.$binaries.log
>     date # timestamp
>   done
> done
Switched to branch 'jca/947-new-ants-binaries'
Your branch is up to date with 'origin/jca/947-new-ants-binaries'.

----------jca/947-new-ants-binaries bin.new-ants-binaries.linux-----------
total 246124
-rwxr-xr-x 1 kousu kousu 28630112 Apr 22 16:11 isct_antsApplyTransforms
-rwxr-xr-x 1 kousu kousu 44594624 Apr 22 16:11 isct_antsRegistration
-rwxr-xr-x 1 kousu kousu 21435952 Apr 22 16:11 isct_antsSliceRegularizedRegistration
-rwxr-xr-x 1 kousu kousu 19880216 Apr 22 16:11 isct_ComposeMultiTransform
-rwxr-xr-x 1 kousu kousu 10439872 Mar  2  2016 isct_dice_coefficient
-rwxr-xr-x 1 kousu kousu 20504754 Jul  5  2016 isct_propseg
-rwxr-xr-x 1 kousu kousu 53186384 Apr  9  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 53331105 Dec  4  2018 isct_train_svm
Sun Apr 26 18:34:40 EDT 2020
    100 X 0.0004752904167741434 -1.1970542445283172e-05 -1.1970542445283172e-05 -1.1970542445283172e-05 -0.12966427418026819 -0.12966427418026819 -0.12966427418026819
Sun Apr 26 18:54:16 EDT 2020

----------jca/947-new-ants-binaries bin.new-ctrDetect-binaries.linux-----------
total 246112
-rwxr-xr-x 1 kousu kousu 28630112 Apr 22 16:11 isct_antsApplyTransforms
-rwxr-xr-x 1 kousu kousu 44594624 Apr 22 16:11 isct_antsRegistration
-rwxr-xr-x 1 kousu kousu 21435952 Apr 22 16:11 isct_antsSliceRegularizedRegistration
-rwxr-xr-x 1 kousu kousu 19880216 Apr 22 16:11 isct_ComposeMultiTransform
-rwxr-xr-x 1 kousu kousu 10439872 Mar  2  2016 isct_dice_coefficient
-rwxr-xr-x 1 kousu kousu 20504754 Jul  5  2016 isct_propseg
-rwxr-xr-x 1 kousu kousu 53172736 Apr 19  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 53331105 Apr 19  2017 isct_train_svm
Sun Apr 26 18:54:16 EDT 2020
    100 X 0.0004752904167741434 -1.1970542445283172e-05 -1.1970542445283172e-05 -1.1970542445283172e-05 -0.12966427418026819 -0.12966427418026819 -0.12966427418026819
Sun Apr 26 19:14:01 EDT 2020
Switched to branch 'ko/new-ctrDetect-binaries'
Your branch is up to date with 'origin/ko/new-ctrDetect-binaries'.

----------ko/new-ctrDetect-binaries bin.new-ants-binaries.linux-----------
total 246124
-rwxr-xr-x 1 kousu kousu 28630112 Apr 22 16:11 isct_antsApplyTransforms
-rwxr-xr-x 1 kousu kousu 44594624 Apr 22 16:11 isct_antsRegistration
-rwxr-xr-x 1 kousu kousu 21435952 Apr 22 16:11 isct_antsSliceRegularizedRegistration
-rwxr-xr-x 1 kousu kousu 19880216 Apr 22 16:11 isct_ComposeMultiTransform
-rwxr-xr-x 1 kousu kousu 10439872 Mar  2  2016 isct_dice_coefficient
-rwxr-xr-x 1 kousu kousu 20504754 Jul  5  2016 isct_propseg
-rwxr-xr-x 1 kousu kousu 53186384 Apr  9  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 53331105 Dec  4  2018 isct_train_svm
Sun Apr 26 19:14:01 EDT 2020
    100 X 0.0004752904167741434 -1.1970542445283172e-05 -1.1970542445283172e-05 -1.1970542445283172e-05 -0.12966427418026819 -0.12966427418026819 -0.12966427418026819
Sun Apr 26 19:33:18 EDT 2020

----------ko/new-ctrDetect-binaries bin.new-ctrDetect-binaries.linux-----------
total 246112
-rwxr-xr-x 1 kousu kousu 28630112 Apr 22 16:11 isct_antsApplyTransforms
-rwxr-xr-x 1 kousu kousu 44594624 Apr 22 16:11 isct_antsRegistration
-rwxr-xr-x 1 kousu kousu 21435952 Apr 22 16:11 isct_antsSliceRegularizedRegistration
-rwxr-xr-x 1 kousu kousu 19880216 Apr 22 16:11 isct_ComposeMultiTransform
-rwxr-xr-x 1 kousu kousu 10439872 Mar  2  2016 isct_dice_coefficient
-rwxr-xr-x 1 kousu kousu 20504754 Jul  5  2016 isct_propseg
-rwxr-xr-x 1 kousu kousu 53172736 Apr 19  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 53331105 Apr 19  2017 isct_train_svm
Sun Apr 26 19:33:18 EDT 2020
    100 X 0.0004752904167741434 -1.1970542445283172e-05 -1.1970542445283172e-05 -1.1970542445283172e-05 -0.12966427418026819 -0.12966427418026819 -0.12966427418026819
Sun Apr 26 19:52:26 EDT 2020
```

</details>

The values  and average runtimes for sct_dmri_moco::`dmri_test1` match the (final case of the) previous PR (https://github.com/neuropoly/spinalcordtoolbox/pull/2642#issuecomment-619385622), so changing `ctrDetect` didn't affect this test at all.

Runtimes per trial:

* 11.76s
* 11.85s
* 11.57s
* 11.48s

I'm re-running this test with `dmri_test2` because that's the one that failed in CI, so maybe it's more unstable for some reason? Maybe it was actually more unstable in #2642 and I just missed that?

kousu commented:
On my laptop, the `dmri_test2` values are also completel consistent and as expected.

<details><summary>`dmri_test2` log</summary>

```
$ for branch in jca/947-new-ants-binaries ko/new-ctrDetect-binaries; do
>   git checkout $branch || break
>   for binaries in bin.new-ants-binaries.linux bin.new-ctrDetect-binaries.linux; do
>     ln -sfT $binaries bin
>     echo
>     echo "----------$branch $binaries-----------"
>     ls -l bin/ # verify we're using the expected binaries
>     date # timestamp
>     for i in `seq 20`; do
>       echo `(cd sct_testing_data; sct_dmri_moco -i dmri/dmri.nii.gz -bvec dmri/bvecs.txt -g 3 -m dmri/mask.nii -x nn -ofolder dmri_test2 -r 1 >/dev/null 2>&1; cat dmri_test2/moco_params.tsv | cut -f 1)`
>     done |
>     sort | uniq -c | tee $(basename $branch)-bin.$binaries.log
>     date # timestamp
>   done
> done
Switched to branch 'jca/947-new-ants-binaries'
Your branch is up to date with 'origin/jca/947-new-ants-binaries'.

----------jca/947-new-ants-binaries bin.new-ants-binaries.linux-----------
total 246124
-rwxr-xr-x 1 kousu kousu 28630112 Apr 22 16:11 isct_antsApplyTransforms
-rwxr-xr-x 1 kousu kousu 44594624 Apr 22 16:11 isct_antsRegistration
-rwxr-xr-x 1 kousu kousu 21435952 Apr 22 16:11 isct_antsSliceRegularizedRegistration
-rwxr-xr-x 1 kousu kousu 19880216 Apr 22 16:11 isct_ComposeMultiTransform
-rwxr-xr-x 1 kousu kousu 10439872 Mar  2  2016 isct_dice_coefficient
-rwxr-xr-x 1 kousu kousu 20504754 Jul  5  2016 isct_propseg
-rwxr-xr-x 1 kousu kousu 53186384 Apr  9  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 53331105 Dec  4  2018 isct_train_svm
Sun Apr 26 19:53:04 EDT 2020

     20 X -0.02831688588348129 0.010687867678295859 0.010687867678295859 0.010687867678295859 -0.09037059326262416 -0.09037059326262416 -0.09037059326262416
Sun Apr 26 19:53:51 EDT 2020

----------jca/947-new-ants-binaries bin.new-ctrDetect-binaries.linux-----------
total 246112
-rwxr-xr-x 1 kousu kousu 28630112 Apr 22 16:11 isct_antsApplyTransforms
-rwxr-xr-x 1 kousu kousu 44594624 Apr 22 16:11 isct_antsRegistration
-rwxr-xr-x 1 kousu kousu 21435952 Apr 22 16:11 isct_antsSliceRegularizedRegistration
-rwxr-xr-x 1 kousu kousu 19880216 Apr 22 16:11 isct_ComposeMultiTransform
-rwxr-xr-x 1 kousu kousu 10439872 Mar  2  2016 isct_dice_coefficient
-rwxr-xr-x 1 kousu kousu 20504754 Jul  5  2016 isct_propseg
-rwxr-xr-x 1 kousu kousu 53172736 Apr 19  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 53331105 Apr 19  2017 isct_train_svm
Sun Apr 26 19:53:51 EDT 2020
     20 X -0.02831688588348129 0.010687867678295859 0.010687867678295859 0.010687867678295859 -0.09037059326262416 -0.09037059326262416 -0.09037059326262416
Sun Apr 26 19:54:36 EDT 2020
Switched to branch 'ko/new-ctrDetect-binaries'
Your branch is up to date with 'origin/ko/new-ctrDetect-binaries'.

----------ko/new-ctrDetect-binaries bin.new-ants-binaries.linux-----------
total 246124
-rwxr-xr-x 1 kousu kousu 28630112 Apr 22 16:11 isct_antsApplyTransforms
-rwxr-xr-x 1 kousu kousu 44594624 Apr 22 16:11 isct_antsRegistration
-rwxr-xr-x 1 kousu kousu 21435952 Apr 22 16:11 isct_antsSliceRegularizedRegistration
-rwxr-xr-x 1 kousu kousu 19880216 Apr 22 16:11 isct_ComposeMultiTransform
-rwxr-xr-x 1 kousu kousu 10439872 Mar  2  2016 isct_dice_coefficient
-rwxr-xr-x 1 kousu kousu 20504754 Jul  5  2016 isct_propseg
-rwxr-xr-x 1 kousu kousu 53186384 Apr  9  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 53331105 Dec  4  2018 isct_train_svm
Sun Apr 26 19:54:37 EDT 2020
     20 X -0.02831688588348129 0.010687867678295859 0.010687867678295859 0.010687867678295859 -0.09037059326262416 -0.09037059326262416 -0.09037059326262416
Sun Apr 26 19:55:22 EDT 2020

----------ko/new-ctrDetect-binaries bin.new-ctrDetect-binaries.linux-----------
total 246112
-rwxr-xr-x 1 kousu kousu 28630112 Apr 22 16:11 isct_antsApplyTransforms
-rwxr-xr-x 1 kousu kousu 44594624 Apr 22 16:11 isct_antsRegistration
-rwxr-xr-x 1 kousu kousu 21435952 Apr 22 16:11 isct_antsSliceRegularizedRegistration
-rwxr-xr-x 1 kousu kousu 19880216 Apr 22 16:11 isct_ComposeMultiTransform
-rwxr-xr-x 1 kousu kousu 10439872 Mar  2  2016 isct_dice_coefficient
-rwxr-xr-x 1 kousu kousu 20504754 Jul  5  2016 isct_propseg
-rwxr-xr-x 1 kousu kousu 53172736 Apr 19  2017 isct_spine_detect
-rwxr-xr-x 1 kousu kousu 53331105 Apr 19  2017 isct_train_svm
Sun Apr 26 19:55:22 EDT 2020
     20 X -0.02831688588348129 0.010687867678295859 0.010687867678295859 0.010687867678295859 -0.09037059326262416 -0.09037059326262416 -0.09037059326262416
Sun Apr 26 19:56:10 EDT 2020
```

</details>

kousu commented:
I (temporarily) replaced the Travis script with my testing script that collates the outputs. Here's the first three runs:

* Linux:
  1. https://api.travis-ci.org/v3/job/679896011/log.txt
    * 100% `0.008491259664160821 0.002902067060777824 0.002902067060777824 0.002902067060777824 -0.014405996135095477 -0.014405996135095477 -0.014405996135095477`
* OS X:
  1. https://api.travis-ci.org/v3/job/679896012/log.txt
    * 90% `0.008491259664160821 0.002902067060777824 0.002902067060777824 0.002902067060777824 -0.014405996135095477 -0.014405996135095477 -0.014405996135095477`
    * 10% `0.008491259664160821 0.0029023974357773596 0.0029023974357773596 0.0029023974357773596 -0.014405747222058807 -0.014405747222058807 -0.014405747222058807`
  2. https://api.travis-ci.org/v3/job/679896013/log.txt 89%
    * 89% `0.008491259664160821 0.002902067060777824 0.002902067060777824 0.002902067060777824 -0.014405996135095477 -0.014405996135095477 -0.014405996135095477`
    * 11% `0.008491259664160821 0.0029023974357773596 0.0029023974357773596 0.0029023974357773596 -0.014405747222058807 -0.014405747222058807 -0.014405747222058807`

kousu commented:
Oh nevermind. The instability in `dmri_test2` has nothing to do with `ctrDetect`. I ran the same hacked up CI script on #2642 to get https://travis-ci.org/github/neuropoly/spinalcordtoolbox/builds/679901666 which has the same ~10% weird behaviour:

* Linux:
  1. https://api.travis-ci.org/v3/job/679901667/log.txt
    * 100% `0.008491259664160821 0.002902067060777824 0.002902067060777824 0.002902067060777824 -0.014405996135095477 -0.014405996135095477 -0.014405996135095477` (same)
* OS X:
  1. https://api.travis-ci.org/v3/job/679901668/log.txt
    * died early as before: `Motion parameters do not match:   results: [0.008491259664160821, 0.00290239743577736, 0.00290239743577736, 0.00290239743577736, -0.014405747222058809, -0.014405747222058809, -0.014405747222058809]  ground truth: [0.008491259664160821, 0.0029020670607778237, 0.0029020670607778237, 0.0029020670607778237, -0.014405996135095476, -0.014405996135095476, -0.014405996135095476]`
  2. https://api.travis-ci.org/v3/job/679901669/log.txt
    * 94% `0.008491259664160821 0.002902067060777824 0.002902067060777824 0.002902067060777824 -0.014405996135095477 -0.014405996135095477 -0.014405996135095477`
    * 6% `0.008491259664160821 0.0029023974357773596 0.0029023974357773596 0.0029023974357773596 -0.014405747222058807 -0.014405747222058807 -0.014405747222058807`

kousu commented:
I feel good about merging this as is.


* [ ]  Vet that this works right on centos6
  * [ ] With `./install_sct`
  * [ ] With deleting bin/ and letting the auto-downloader kick in again

Here are some follow up work items:

* [ ] Manually eyeball the intermediate registration results from the various tests
  * Is there a way we can turn "eyeballing" into "unit testing"?
* [ ] Performance testing on a big multi-CPU computer
* [ ] Clean up `install/`
  * [ ] `git rm install/{compile_ants.py,copy_ants_binaries.py}`
  * [ ] `git rm install/make_binaries.sh`
  * [ ] `git rm install/compile_external.py` ?
  * [ ] `git rm install/compress_binaries.sh` ?
  * [ ] `git rm install/install_dependencies_with_workarounds.py` ?
  * I want to do this in a followup PR

kousu commented:
Despite this downgrading the binaries, I want to merge it for consistency's sake -- particularly, so that the build scripts for #2668 can refer to upstream directly instead of us having to keep a weird pseudo-mirror that no one else has just of those.

It turned out that all the inconsistencies above were just problems in #2642, that are fixed now, so whatever downgrade this is it doesn't change the numbers appreciably.

kousu commented:
It's not going to happen this week or next, sorry, I'm too busy with other projects.

joshuacwnewton commented:
I've cleaned up this PR a bit so that it is easier to tackle, but I want to summarize the changes I made.

![image](https://user-images.githubusercontent.com/16181459/167141248-3d450255-0eb0-4b6e-9903-3d2d47b1a47d.png)

* This PR originally had 8 commits. However, 7/8 of those commits were actually just duplicates of PR [#2642](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2642), which had already been merged. The one remaining (unique) commit was: https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/9d097853684e35998a7a496d4cedc125b957f8ca.
* This PR originally had a lot of discussion about numerical stability. However, as @kousu mentioned in [this comment](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2666#issuecomment-623800953), the discussion is only relevant to the 7 duplicate commits that were already merged, and not to the 1 commit that remains.
* So, I discarded the 7 duplicate commits, then rebased the changes from that 1 commit onto master.

Now to focus on that 1 remaining commit (https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/9d097853684e35998a7a496d4cedc125b957f8ca):

* The 1 remaining commit (https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/9d097853684e35998a7a496d4cedc125b957f8ca) tried to change the OSF links in order to update the ctrDetect binaries. 
* But, PR [#2821](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2821) _also_ changed the OSF links (https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2821/commits/18b0fe64ae3c7664dc6d521b7d5d3da5e87b791d), but with the purpose of updating the ANTs binaries instead!
* So, there are actually 3 total links:

1. `binaries_linux`
    * Old link (i.e. previous master):  https://osf.io/mka78/?action=download
    * Current link (i.e. present master): https://osf.io/cs6zt/?action=download
    * Proposed link: https://osf.io/admxy/?action=download
2. `binaries_osx`
    * Old link (i.e. previous master):  https://osf.io/dn67h/?action=download
    * Current link (i.e. present master): https://osf.io/874cy?action=download
    * Proposed link: https://osf.io/7vx5g/?action=download

Because of this, I believe what needs to be done is to create one final OSF link that merges the updated ANTs and ctrDetect binaries from both this PR and https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2821 into one combined package.

joshuacwnewton commented:
To explain the 1 remaining commit:

* The commit (https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/9d097853684e35998a7a496d4cedc125b957f8ca) tried to change the OSF links in order to update the ctrDetect binaries. 
* But, since then, merged PR [#2821](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2821) had _also_ changed the OSF links (https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2821/commits/18b0fe64ae3c7664dc6d521b7d5d3da5e87b791d), but with the purpose of updating the ANTs binaries instead!
* Additionally, the currently-open PR [#2668](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2668) _also_ updates the OSF binary links.
* So, there are actually _5_ sets of links:

1. `binaries_linux`
    * Old link (i.e. previous master): [`2020-04-20`](https://osf.io/mka78/?action=download)
    * Proposed link 1 (i.e. this PR): [`2020-04-21`](https://osf.io/admxy/?action=download)
    * Proposed link 2 (i.e. PR #[2667](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2667)): [`2020-04-22`](https://osf.io/xhr4v/?action=download)
    * Proposed link 3 (i.e. PR #[2668](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2668)): [`kousu/spinalcordtoolbox-binaries`](https://github.com/kousu/spinalcordtoolbox-binaries/releases/download/refs%2Fpull%2F3%2Fmerge-85120650/spinalcordtoolbox-binaries_linux.tar.gz)
    * Current link (i.e. present master): [`2020-08-01`](https://osf.io/cs6zt/?action=download)

2. `binaries_osx`
    * Old link (i.e. previous master): [`2020-04-20`](https://osf.io/dn67h/?action=download)
    * Proposed link 1 (i.e. this PR): [`2020-04-21`](https://osf.io/7vx5g/?action=download)
    * Proposed link 2 (i.e. PR #[2667](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2667)): [`2020-04-22`](https://osf.io/eg2m5/?action=download)
    * Proposed link 3 (i.e. PR  #[2668](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2668)): [`kousu/spinalcordtoolbox-binaries`](https://github.com/kousu/spinalcordtoolbox-binaries/releases/download/refs%2Fpull%2F3%2Fmerge-85120650/spinalcordtoolbox-binaries_osx.tar.gz)
    * Current link (i.e. present master): [`2020-08-01`](https://osf.io/874cy?action=download)

So, I believe what need to do is create 1 final OSF package that combines the changes from #2666, #2667, #2668, and #2821.

joshuacwnewton commented:
I've cleaned up this PR a bit so that it is easier to understand, and I want to summarize the changes I made.

![image](https://user-images.githubusercontent.com/16181459/167141248-3d450255-0eb0-4b6e-9903-3d2d47b1a47d.png)

* This PR originally had 8 commits. However, 7/8 of those commits were actually just duplicates of PR [#2642](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2642), which had already been merged. The one remaining (unique) commit was: https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/9d097853684e35998a7a496d4cedc125b957f8ca.
* This PR originally had a lot of discussion about numerical stability. However, as @kousu mentioned in [this comment](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2666#issuecomment-623800953), the discussion was only relevant to the 7 duplicate commits that were already merged, and not to the 1 commit that remains.

So, I discarded the 7 duplicate commits, then rebased the changes from that 1 commit onto master.

joshuacwnewton commented:
Closing this PR as it is likely to be superseded by #3788/#2668.
